File: JumpGameII.py
import sys

class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        steps = [0 for i in range(len(A))]
        #steps[len(A) - 1] = 1 if A[len(A) - 1] > 0 else 0
        for i in range(len(A) - 2, -1, -1):
            min = sys.maxsize
            for j in range(A[i]):
                if i + j + 1 < len(A):
                    if steps[i + j + 1] + 1 < min:
                        min = steps[i + j + 1] + 1
                        print("Updateing " + str(i) + " with " + str(i + j + 1))
            steps[i] = min
            print("i becomes: " + str(steps[i]))
            print(steps)
        return steps[0]

sol = Solution()
print(sol.jump([2, 3, 1, 1, 4]))

File: JumpGameII.cc
class Solution {
public:
    int jump(int A[], int n) {
        if (n < 2) {
            return 0;
        }
        int range_start = 0, range_end = 0;
        int counter = 0;
        while (range_end < n) {
            counter++;
            int current_max = 0;
            for(int i = range_start; i <= range_end; i++) {
                if(A[i] + i >= n - 1) {
                    return counter;
                }

                if (A[i] + i > current_max) {
                    current_max = A[i] + i;
                }
            }

            range_start = range_end + 1;
            range_end = current_max; // > range_end + 1 ? current_max : range_end + 1;
        }

        return counter;
    }
};
